src/adapter/src/coord/command_handler.rs

```diff
@@ -261,6 +261,30 @@ impl Coordinator {
                 let mut session_defaults = BTreeMap::new();
                 let system_config = self.catalog().state().system_config();
 
+                // Validate network policies for external users. Internal users
+                // can only connect on the internal interafces(internal http/
+                // pgwire). It is up to the person deploying the system to
+                // ensure these internal interfaces are well secured.
+                if !user.is_internal() {
+                    let default_policy = NetworkPolicy::new(system_config.default_network_policy());
+                    if let Some(ip) = client_ip {
+                        match default_policy.validate(&ip) {
+                            Ok(_) => {}
+                            Err(e) => {
+                                let _ = tx
+                                    .send(Err(AdapterError::NetworkPolicyDenied(e)))
+                                    .inspect_err(|e| {
+                                        error!("Failed sending session denied error: {:?}", e)
+                                    });
+                                // The connection was deinied before internal
+                                // state for the connection was setup. We should
+                                // not have to call handle_terminate.
+                                return;
```

Comment:
`handle_startup_inner(...)` creates a temporary schema for this connection which we need to cleanup. I think the cleanest thing to do here is move this entire check into `handle_startup_inner(...)` since if that returns an error we already do the cleanup, looks like we should have everything we need in that method?

Reply:
oh great catch!